### PR TITLE
S(Q, w) table checks

### LINKED ISSUE #26637: S(Q,omega) table checks
https://docs.mantidproject.org/nightly/algorithms/SofQWNormalisedPolygon-v1.html#algm-sofqwnormalisedpolygon should have a validate inputs, which makes sure that the table is of the correct form (int, double, double). We should also add unit tests to check that it catches invalid tables. 

We should also check that the table and workspace have the same number of detectors.

The following script should  create an invalid table and should be caught at the validate inputs stage. If we run the algorithm from the GUI then it would give the red star next to the table. 

```
import numpy as np
# Create Table to hold detector angular range
IRIS_PG_Det = CreateEmptyTableWorkspace()

# Calculate range for PG detectors on IRIS
no_det = int(51)
det_start = int(3)
detIDList = [i+det_start for i in range(no_det)]
angle_start = np.radians(25.)
angle_end = np.radians(160.)
angles = np.linspace(angle_start, angle_end, no_det+1, endpoint = True)
min = angles[:-1]
max = angles[1:]

# Add some columns, Recognized types are: int,float,double,bool,str,V3D,long64
IRIS_PG_Det.addColumn(type="int",name="Detector ID")
IRIS_PG_Det.addColumn(type="int",name="Min two theta")
IRIS_PG_Det.addColumn(type="double",name="Max two theta")

# Populate the columns for three detectors
for i,j in enumerate(detIDList):
    IRIS_PG_Det.addRow ([j, i, max[i] ] )
```

### PR DESCRIPTION
**Pull request by ciaranightingale**
**Description of work.**
https://docs.mantidproject.org/nightly/algorithms/SofQWNormalisedPolygon-v1.html#algm-sofqwnormalisedpolygon now has validateInputs function to check that the table is of the correct form (int, double, double), that the input workspace is of the correct type and that the table and workspace have the same number of detectors. 

**To test:**

The following script should create an invalid table and should be caught at the validate inputs stage. If we run the algorithm from the GUI then it would give the red star next to the table.
```
import numpy as np

#Create Table to hold detector angular range
IRIS_PG_Det = CreateEmptyTableWorkspace()

#Calculate range for PG detectors on IRIS
no_det = int(51)
det_start = int(3)
detIDList = [i+det_start for i in range(no_det)]
angle_start = np.radians(25.)
angle_end = np.radians(160.)
angles = np.linspace(angle_start, angle_end, no_det+1, endpoint = True)
min = angles[:-1]
max = angles[1:]

#Add some columns, Recognized types are: int,float,double,bool,str,V3D,long64
IRIS_PG_Det.addColumn(type="int",name="Detector ID")
IRIS_PG_Det.addColumn(type="int",name="Min two theta")
IRIS_PG_Det.addColumn(type="double",name="Max two theta")

#Populate the columns for three detectors
for i,j in enumerate(detIDList):
    IRIS_PG_Det.addRow ([j, i, max[i] ] )
```

Fixes #26637 . 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Algorithms/test/SofQWNormalisedPolygonTest.h [hunk 1]
```diff
@@ -358,6 +358,106 @@ class SofQWNormalisedPolygonTest : public CxxTest::TestSuite {
...
+    alg->initialize();
+    alg->setProperty("InputWorkspace", inWS);
+    alg->setProperty("EMode", "Indirect");
+    alg->setProperty("QAxisBinning", qBinParams);
+    alg->setProperty("DetectorTwoThetaRanges", twoThetaRanges);
+    return alg;
+  }
+
+  /** Create a table workspace
+   * @param types A vector of strings of the data types of the columns
```
[the remark this answers is not in the document]
Comment (AnthonyLim23) [reply Framework/Algorithms/test/SofQWNormalisedPolygonTest.h]: would the name `dataType` be clearer?

[the remark this answers is not in the document]
Comment (ciaranightingale) [reply Framework/Algorithms/test/SofQWNormalisedPolygonTest.h]: already done!

**Comment by AnthonyLim23** [comment]
@ciaranightingale you have left in the diff comments as part of your last commit.

**Review by AnthonyLim23 — DISMISSED** [review]
:shipit: once builds pass

**Review by AnthonyLim23 — DISMISSED** [review]
:shipit:

File: Framework/Algorithms/src/SofQWNormalisedPolygon.cpp [hunk 2]
```diff
@@ -311,12 +311,56 @@ void SofQWNormalisedPolygon::init() {
...
+std::map<std::string, std::string> SofQWNormalisedPolygon::validateInputs() {
+  std::map<std::string, std::string> result;
+  API::MatrixWorkspace_const_sptr inputWS = getProperty("InputWorkspace");
+  if (!inputWS) {
+    result["InputWorkspace"] = "InputWorkspace is of Incorrect type. Please "
+                               "provide a MatrixWorkspace as the "
+                               "InputWorkspace";
+  }
+  TableWorkspace_sptr tableWS = getProperty("DetectorTwoThetaRanges");
+  if (tableWS) {
```
Comment (DanNixon) [inline Framework/Algorithms/src/SofQWNormalisedPolygon.cpp]: This block can be simplified slightly.

The `return` in the first `if` can be removed and the following `if`s converted to `else if`.

It only makes sense to assign `result["DetectorTwoThetaRanges"]` once.

File: docs/source/release/v4.2.0/indirect_geometry.rst [hunk 3]
```diff
@@ -15,6 +15,7 @@ Improved
 - Instrument definition is improved for IN16B to have the physical detector on the correct side of the beam axis, and different analyser focus for single detectors.
 - :ref:`LoadILLIndirect <algm-LoadILLIndirect>` is extended to support also the configurations with the first tube angle at 33.1 degrees.
 - :ref:`IndirectILLEnergyTransfer <algm-IndirectILLEnergyTransfer>` now offers the possibility to enable or disable the detector grouping both for Doppler and BATS modes. By default the pixels will be grouped tube by tube as before.
+- :ref:`SofQWNormalisedPolygon <algm-SofQWNormalisedPolygon>` now has a function to check the the input properties are valid.
```
[the remark this answers is not in the document]
Comment (DanNixon) [reply docs/source/release/v4.2.0/indirect_geometry.rst]: Was this done? It still mentions "function".

Perhaps something like "... now checks input properties are valid"?

**Review by DanNixon — CHANGES_REQUESTED** [review]
Looks good :+1:, just a couple of comments.

**Comment by AnthonyLim23** [comment]
the error messages have changed.

**Comment by AnthonyLim23** [comment]
@ciaranightingale you still need to push the changes

**Comment by DanNixon** [comment]
Unless there is some actual change to be made can the last two commits please be removed. They simply add and remove a back quote.

File: docs/source/release/v4.2.0/indirect_geometry.rst [hunk 4]
```diff
@@ -15,6 +15,7 @@ Improved
 - Instrument definition is improved for IN16B to have the physical detector on the correct side of the beam axis, and different analyser focus for single detectors.
 - :ref:`LoadILLIndirect <algm-LoadILLIndirect>` is extended to support also the configurations with the first tube angle at 33.1 degrees.
 - :ref:`IndirectILLEnergyTransfer <algm-IndirectILLEnergyTransfer>` now offers the possibility to enable or disable the detector grouping both for Doppler and BATS modes. By default the pixels will be grouped tube by tube as before.
+- :ref:`SofQWNormalisedPolygon <algm-SofQWNormalisedPolygon>`now checks input properties are valid.
```
Comment (DanNixon) [inline docs/source/release/v4.2.0/indirect_geometry.rst]:
```suggestion
- :ref:`SofQWNormalisedPolygon <algm-SofQWNormalisedPolygon>` now checks input properties are valid.
```

**Review by DanNixon — CHANGES_REQUESTED** [review]

**Review by DanNixon — APPROVED** [review]

**Pull request merged**